code/iaas/logic/src/main/java/io/cattle/platform/process/containerevent/ContainerEventCreate.java

```diff
@@ -0,0 +1,207 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.apache.commons.lang3.BooleanUtils;
+
+import com.netflix.config.DynamicBooleanProperty;
+
+@Named
+public class ContainerEventCreate extends AbstractDefaultProcessHandler {
+
+    private static final DynamicBooleanProperty MANAGE_NONRANCHER_CONTAIERS = ArchaiusUtil
+            .getBoolean("manage.nonrancher.containers");
+    private static final String INSPECT_ENV = "Env";
+    private static final String INSPECT_NAME = "Name";
+    private static final String FIELD_DOCKER_INSPECT = "dockerInspect";
+    private static final String INSPECT_CONFIG = "Config";
+    private static final String INSPECT_IMAGE = "Image";
+    private static final String IMAGE_PREFIX = "docker:";
+    private static final String IMAGE_KIND_PATTERN = "^(sim|docker):.*";
+    private static final String RANCHER_UUID = "RANCHER_UUID=";
+    private static final String RANCHER_NETWORK = "RANCHER_NETWORK=";
+
+    @Inject
+    StorageService storageService;
+
+    @Inject
+    NetworkDao networkDao;
+
+    @Inject
+    AccountDao accountDao;
+
+    @Inject
+    LockManager lockManager;
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        if ( !MANAGE_NONRANCHER_CONTAIERS.get() ) {
+            return null;
+        }
+
+        ContainerEvent event = (ContainerEvent)state.getResource();
+        Map<String, Object> inspect = (Map<String, Object>)DataUtils.getFields(event).get(FIELD_DOCKER_INSPECT);
+
+        String rancherUuid = getRancherUuidLabel(inspect);
+        if ( rancherUuid != null ) {
+            return null;
+        }
+
+        scheduleInstance(event, inspect);
+
+        return null;
+    }
+
+    void scheduleInstance(final ContainerEvent event, final Map<String, Object> inspect) {
+        final Long accountId = event.getAccountId();
+        final String externalId = event.getExternalId();
+
+        lockManager.lock(new ContainerEventInstanceLock(accountId, externalId), new LockCallbackNoReturn() {
+            @Override
+            public void doWithLockNoResult() {
+                // TODO INDEX ON EXTERNAL ID
+                Instance instance = objectManager.findAny(Instance.class, INSTANCE.ACCOUNT_ID, accountId,
+                        INSTANCE.EXTERNAL_ID, externalId);
+                if ( instance == null ) {
+                    instance = objectManager.newRecord(Instance.class);
+                    instance.setKind(InstanceConstants.KIND_CONTAINER);
+                    instance.setAccountId(accountId);
+                    instance.setExternalId(externalId);
+                    setName(inspect, instance);
+                    setNetwork(accountId, inspect, instance);
+                    setImage(inspect, instance);
+                    setHost(event, instance);
```

Comment:
@cjellick is there a particular reason why we schedule instance restart on the same host instead of delegating this decision to allocators? I would set it only in case that user set requestedHostId on instance during its original start. Otherwise limiting it to the same host can cause failure situation to repeat - if the host is in some failure condition for example.